src/interface/control_scriptbox.cpp

```diff
@@ -0,0 +1,496 @@
+      {
+        m_hovered_parameter = &parameter;
+        break;
+      }
+    }
+    return true;
+  }
+
+  return ControlTextbox::on_mouse_motion(motion);
+}
+
+bool
+ControlScriptbox::on_mouse_wheel(const SDL_MouseWheelEvent& wheel)
+{
+  if (!m_has_focus)
+    return false;
+
+  if (m_suggestions.empty())
+    return ControlTextbox::on_mouse_wheel(wheel);
+
+  return m_suggestions_scrollbar->on_mouse_wheel(wheel);
+}
+
+bool
+ControlScriptbox::on_key_down(const SDL_KeyboardEvent& key)
+{
+  if (m_suggestions.empty())
+    return ControlTextbox::on_key_down(key);
+
+  switch (key.keysym.sym)
+  {
+    case SDLK_UP:
+    {
+      if (m_selected_suggestion > 0)
+      {
+        m_selected_suggestion--;
+
+        const float line_pos = m_selected_suggestion * Resources::control_font->get_height() + TEXT_Y_OFFSET;
+        if (line_pos - m_suggestions_offset < 0.f)
+          m_suggestions_offset = line_pos;
+
+        update_description();
+        return true;
+      }
+      else
+      {
+        m_suggestions.clear();
+      }
+      break;
+    }
+
+    case SDLK_DOWN:
+    {
+      if (m_selected_suggestion < m_suggestions.size() - 1)
+      {
+        m_selected_suggestion++;
+
+        const float line_pos = (m_selected_suggestion + 1) * Resources::control_font->get_height() + TEXT_Y_OFFSET;
+        if (line_pos - m_suggestions_offset > m_suggestions_rect.get_height())
+          m_suggestions_offset = line_pos - m_suggestions_rect.get_height();
+
+        update_description();
+        return true;
+      }
+      else
+      {
+        m_suggestions.clear();
+      }
+      break;
+    }
+
+    case SDLK_TAB:
+    case SDLK_RETURN: // ENTER
+    {
+      autocomplete();
+      return true;
+    }
+  }
+
+  return ControlTextbox::on_key_down(key);
+}
+
+void
+ControlScriptbox::update_description()
+{
+  m_suggestion_description.clear();
+  m_function_parameters.clear();
+  m_hovered_parameter = nullptr;
+
+  if (m_suggestions.empty())
+    return;
+
+  const squirrel::Suggestion& suggestion = m_suggestions.at(m_selected_suggestion);
+  if (!suggestion.reference) return;
+
+  const float max_width = (static_cast<float>(SCREEN_WIDTH) - m_suggestions_rect.get_right() -
+                           SUGGESTION_DESCRIPTION_RECT_SPACING);
+
+  using namespace squirrel;
+
+  std::vector<std::string> desc_lines;
+  switch (suggestion.reference->get_type())
+  {
+    case ScriptingObject::Type::CONSTANT:
+    {
+      const auto& con = static_cast<const ScriptingConstant&>(*suggestion.reference);
+
+      desc_lines.push_back(con.type + " " + con.name);
+      desc_lines.push_back(con.description);
+      desc_lines.push_back(con.detailed_description);
+    }
+    break;
+
+    case ScriptingObject::Type::FUNCTION:
+    {
+      const auto& func = static_cast<const ScriptingFunction&>(*suggestion.reference);
+
+      const Vector pos = get_suggestion_description_rect_pos();
+
+      std::stringstream out;
```

Comment:
Maybe you should use `format` for this instead? There is (theoretically, aka unneccesary, yet likely not practically) a small performance loss doing this that we could probably improve.